specs/_features/eip7732/beacon-chain.md

```diff
@@ -0,0 +1,716 @@
+                              execution_engine: ExecutionEngine, verify: bool = True) -> None:
+    # Verify signature
+    if verify: 
+        assert verify_execution_payload_envelope_signature(state, signed_envelope)
+    envelope = signed_envelope.message
+    payload = envelope.payload
+    # Cache latest block header state root
+    previous_state_root = hash_tree_root(state)
+    if state.latest_block_header.state_root == Root():
+        state.latest_block_header.state_root = previous_state_root
+    
+    # Verify consistency with the beacon block
+    assert envelope.beacon_block_root == hash_tree_root(state.latest_block_header)
+
+    # Verify consistency with the committed header
+    committed_header = state.latest_execution_payload_header
+    assert envelope.builder_index == committed_header.builder_index
+    assert committed_header.blob_kzg_commitments_root == hash_tree_root(envelope.blob_kzg_commitments)
+
+    if not envelope.payload_withheld: 
+        # Verify the withdrawals root
+        assert hash_tree_root(payload.withdrawals) == state.latest_withdrawals_root
+
+        # Verify the gas_limit
+        assert committed_header.gas_limit == payload.gas_limit
+
+        assert committed_header.block_hash == payload.block_hash 
+        # Verify consistency of the parent hash with respect to the previous execution payload
+        assert payload.parent_hash == state.latest_block_hash
+        # Verify prev_randao
+        assert payload.prev_randao == get_randao_mix(state, get_current_epoch(state))
+        # Verify timestamp
+        assert payload.timestamp == compute_timestamp_at_slot(state, state.slot)
+        # Verify commitments are under limit
+        assert len(envelope.blob_kzg_commitments) <= MAX_BLOBS_PER_BLOCK
+        # Verify the execution payload is valid
+        versioned_hashes = [kzg_commitment_to_versioned_hash(commitment) 
+                            for commitment in envelope.blob_kzg_commitments]
+        assert execution_engine.verify_and_notify_new_payload(
+            NewPayloadRequest(
+                execution_payload=payload,
+                versioned_hashes=versioned_hashes,
+                parent_beacon_block_root=state.latest_block_header.parent_root,
+            )
+        )
+
+        # Process Electra operations
+        def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -> None:
+            for operation in operations:
+                fn(state, operation)
+
+        for_ops(payload.deposit_requests, process_deposit_request)
+        for_ops(payload.withdrawal_requests, process_withdrawal_request)
+        for_ops(payload, process_consolidation_request)
+
+        # Cache the execution payload header and proposer
+        state.latest_block_hash = payload.block_hash
+        state.latest_full_slot = state.slot
+
+    # Verify the state root
+    if verify: 
+        assert envelope.state_root == hash_tree_root(state)
```

Comment:
Is this check required? Also, `state` is already updated, and is not reverted if verification fails.

Reply:
The `envelope.payload.state_root` is the EL post-state root and used to check that the execution payload processed correctly.

`envelope.state_root` refers to some CL post-state. It is not linked to the execution payload.

Reply:
Hm. Here `envelope.state_root` is the EL post state root used to check that the execution payload processed correctly
`envelope` is `signed_envelope.message`

Reply:
no. the EL post state root is `envelope.payload.state_root`.

`envelope.state_root` is `hash_tree_root(state)` (of the beacon state -- EL does not use htr)

Reply:
I misspoke. I meant the EL post-state root as the state root after processing the execution payload envelope, but that's not accurate. What I intended to say is that we need to check `signed_envelope.message.state_root == hash_tree_root(state)` to ensure that the post-state aligns. This is similar to checking `block.state_root == hash_tree_root(state)` after running the state transition for the beacon state.